File: 0001-0500/0131_partition.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        output = []
        path = []
        self.dfs(s, output, path)
        return output 
    
    def dfs(self, string, output, path):
        if not string:
            output.append(path)
            return
        for i in range(len(string)):
            sub = string[:i+1]
            if self.isPalindrom(sub):
                self.dfs(string[i+1:], output, path + [sub])
                
    def isPalindrom(self, string):
        i = 0
        j = len(string) - 1
        while i < j:
            if string[i] == string[j]:
                i += 1
                j -= 1
            else:
                return False
        return True

# Alternative solution

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        self.result = []
        self.backtrack(s, [])
        return self.result

    def backtrack(self, string, path):
        if not string:
            self.result.append(path)
            return
        for i in range(len(string)):
            substring = string[:i+1]
            if substring == substring[::-1]:
                self.backtrack(string[i+1:], path + [substring])

File: 0001-0500/0131_partition.java
class Solution {
    public List<List<String>> partition(String s) {
        List<List<String>> output = new ArrayList<>();
        backtrack(s, output, new ArrayList<>());
        return output;
    }

    public void backtrack(String s, List<List<String>> output, List<String> path) {
        if (s.length() == 0) {
            output.add(new ArrayList<>(path));
            return;
        }

        for (int i = 0; i < s.length(); i++) {
            String str = s.substring(0, i + 1);
            if (isPalindrome(str)) {
                path.add(str);
                backtrack(s.substring(i + 1), output, path);
                path.remove(path.size() - 1);
            }
        }
    }

    public boolean isPalindrome(String s) {
        for (int i = 0; i < s.length()/2; i++) {
            if (s.charAt(i) != s.charAt(s.length() - 1 - i)) {
                return false;
            }
        }
        return true;
    }
}
